EMFILE: too many open files when attempting to render a large (170000) number of pages (Eleventy 2.0.0-canary.16)

**Describe the bug**
EMFILE: too many open files when attempting to render a site with a large number of pages from a large data set.

When it is triggered, it is triggered via EleventyTemplateError.

The number of files being opened is finite: on a system with raised limits it successfully builds.

I have one 40MB json file, containing an array of about 170000 objects, and one template that generates one page per entry. ([The site](https://github.com/kisaragi-hiu/kemdict/tree/e2627) is supposed to be a static dictionary.)

Eleventy version: 2.0.0-canary.16. The site only builds on 2.0 due to #2360.

This behavior has been documented before in <https://github.com/11ty/eleventy/issues/2226#issuecomment-1120188230>.

Personally, I think this is an issue that can be fixed, as there is no reason why the output files, which depend only on the input file and not on each other, should all be held open at the same time.

**To Reproduce**

Edit: An earlier version said it took an hour to see the error. That is incorrect: I misread my CI logs, and Eleventy only took 3 minutes.
Edit: This has been rewritten to be more exact and to use error messages from the site generated during the reproduction.

Steps to reproduce the behavior:

1. Create a new test folder
2. Generate the test data and put it in `_data/a.json`, containing an array with a large number of objects, with each object having a unique, say, "title" key. I use Emacs for this:

   ```sh
   mkdir -p _data
   cat > x.el <<"HERE"
   (require 'json)
   (let (lst)
     (dotimes (i 200000)
       (push (let ((tmp (make-hash-table :test #'equal)))
               (puthash \"title\" (format \"%s\" i) tmp)
               tmp)
             lst))
     (with-temp-file \"_data/a.json\"
       (insert (json-encode lst))))
   HERE
   emacs --script x.el
   ```

   The JSON looks something like this:

   ```json
   [{"title": "199999"}, {"title": "199998"}, ...]
   ```

3. Create one template to generate pages for each item:

   ```markdown
   ---
   pagination:
     data: a
     size: 1
     alias: item
   permalink: "item/{{ item.title }}/index.html"
   ---
   {{ item.title }}
   ```

3. `npx @11ty/eleventy --quiet` (`--quiet` is to silence the list of rendered files)
4. On a system with higher FD limits, this will succeed.
5. On a more constrained system, an EMFILE error will be raised (from EleventyTemplateError).

   On Linux the FD limit can be lowered for a command like this:

   ```shell
   prlimit --nofile=50000 npx @11ty/eleventy --quiet
   ```

```
$ prlimit --nofile=50000 npx @11ty/eleventy --quiet
[11ty] Problem writing Eleventy templates: (more in DEBUG output)
[11ty] 1. Having trouble writing to "_site/item/199999/index.html" from "./a.njk" (via EleventyTemplateError)
[11ty] 2. EMFILE: too many open files, open '_site/item/49141/index.html' (via Error)
[11ty] 
[11ty] Original error stack trace: Error: EMFILE: too many open files, open '_site/item/49141/index.html'
[11ty] Wrote 0 files in 93.93 seconds (v2.0.0-canary.16)
```

**Expected behavior**
Writing many files from one template should not cause an EMFILE error.

**Environment:**

Where the error is encountered:

- OS and Version: Ubuntu 20.04 (on GitHub Actions), with an FD limit of somewhere between 1000 and 20000
- Eleventy Version: 2.0.0-canary.16

Where the error is not encountered:

- OS and Version: Arch Linux, with an FD limit in the millions
- Eleventy Version: 2.0.0-canary.16

**Additional context**

I can see that in TemplateWriter.js (generateTemplates) there is essentially an async, unbuffered loop to _generateTemplate → generateMapEntry (Template.js) → _write → writeFile (from graceful-fs).

(This is the source of the EMFILE because it's the only place where EleventyTemplateError can be raised.)

This would rapidly exhaust file descriptors if graceful-fs weren't used. But graceful-fs *is* used, and it should have handled this, and I don't understand why it's happening.

Perhaps Eleventy should also queue up file writes?

@kisaragi-hiu How do you build that site?
I have it cloned locally on macOS, and ran <kbd>npm install</kbd> and then <kbd>make src/_data</kbd>, but I get the following error:

> ```sh
> make src/_data
>
> mkdir -p src/_data
> cp moedict-data/dict-revised.json src/_data/dict-moe-revised.json
> cp: moedict-data/dict-revised.json: No such file or directory
> make: *** [src/_data/dict-moe-revised.json] Error 1
> ```

Right, sorry for not formatting the linked site properly. I've created a new branch for testing ([`e2627`](https://github.com/kisaragi-hiu/kemdict/tree/e2627)) that I won't modify further, and I've added build instructions (clone submodules, install Cask, `cask install`, `npm install`, `make`).

Writing out like 170000 output files is the expected result. At least on Linux it seems to be relatively doable: it builds on my PC (with 8GB of RAM), and it even builds on GitHub Actions (Ubuntu) after I raise the FD limit (`sudo sysctl -w fs.file-max=500000; sudo prlimit --nofile=500000 npx @11ty/eleventy`).

I tested on my PC further, using `prlimit` to lower the limit. The builds starts failing at a limit around 80000 --- which is weird.

```
prlimit --nofile=50000 npx "@11ty/eleventy" --input=./src --output=./_site --quiet # fails
prlimit --nofile=75000 npx "@11ty/eleventy" --input=./src --output=./_site --quiet #fails
prlimit --nofile=78125 npx "@11ty/eleventy" --input=./src --output=./_site --quiet #fails
prlimit --nofile=81250 npx "@11ty/eleventy" --input=./src --output=./_site --quiet # succeeds. What?
prlimit --nofile=100000 npx "@11ty/eleventy" --input=./src --output=./_site --quiet # succeeds
```

I also [tried building it on macOS on GitHub Actions](https://github.com/kisaragi-hiu/kemdict/actions/runs/3334612447/jobs/5517750742). The same error happens:

```
▶ Run gmake _site
npx tailwindcss --minify --postcss -i css/src.css -o _site/css/built.css

Done in 613ms.
npx "@11ty/eleventy" --input=./src --output=./_site --quiet
[11ty] Problem writing Eleventy templates: (more in DEBUG output)
[11ty] 1. Having trouble writing to "./_site/word/{[8e40]}/index.html" from "./src/word-pages.njk" (via EleventyTemplateError)
[11ty] 2. EMFILE: too many open files, open './_site/word/趑趄不前/index.html' (via Error)
[11ty] 
[11ty] Original error stack trace: Error: EMFILE: too many open files, open './_site/word/趑趄不前/index.html'
[11ty] Wrote 1 file in 172.34 seconds (v2.0.0-canary.16)
gmake: *** [Makefile:17: _site] Error 1
Error: Process completed with exit code 2.
```

--- a/src/Template.js
+++ b/src/Template.js
@@ -25,7 +25,6 @@ import EleventyBaseError from "./Errors/EleventyBaseError.js";
 import ReservedData from "./Util/ReservedData.js";
 
 const { set: lodashSet, get: lodashGet } = lodash;
-const writeFile = util.promisify(fs.writeFile);
 const fsStat = util.promisify(fs.stat);
 
 const debug = debugUtil("Eleventy:Template");
@@ -748,7 +747,9 @@ class Template extends TemplateContent {
 		let templateBenchmark = this.bench.get("Template Write");
 		templateBenchmark.before();
 
-		await writeFile(outputPath, finalContent);
+		// Note: This deliberately uses the synchronous version to avoid
+		// unbounded concurrency: https://github.com/11ty/eleventy/issues/3271
+		fs.writeFileSync(outputPath, finalContent);
 
 		templateBenchmark.after();
 		this.writeCount++;

